File: fabric-audit-agent/core/dax.js
// Heuristic text rules — may have false positives; intended as hints, not verdicts.
const RULES = [
  { test: /FILTER\s*\(\s*'?[A-Za-z0-9_ ]+'?\s*,/i, pattern: 'filter-whole-table',
    suggestion: 'FILTER() over an entire table is slow — filter a single column (KEEPFILTERS or a boolean column filter) instead.' },
  { test: /(SUMX|AVERAGEX|MAXX|MINX|COUNTX)[\s\S]*?(SUMX|AVERAGEX|MAXX|MINX|COUNTX)/i, pattern: 'nested-iterators',
    suggestion: 'Possible nested iterators (X-functions inside X-functions). If nested, they evaluate row-by-row quadratically — precompute with a VAR or an aggregation table.' },
  { test: /\bCALCULATE\b[\s\S]*\bCALCULATE\b/i, pattern: 'repeated-calculate',
    suggestion: 'Repeated CALCULATE recomputes filter context — hoist shared expressions into VARs.' },
  { test: /\bEARLIER\s*\(/i, pattern: 'earlier',
    suggestion: 'EARLIER() signals legacy row-context nesting — refactor with VARs.' },
  { test: /(?<![/:])[^/]\/[^/]/, pattern: 'raw-division',
    suggestion: 'Use DIVIDE(numerator, denominator) instead of "/" for safe, efficient division.' },
];

/**
 * Analyze a DAX measure for performance anti-patterns. Pure (rule-based text analysis).
 * @param {string} measureText  the DAX expression
 * @param {{ durationMs?:number }} [stats]
 * @returns {Array<{ pattern:string, suggestion:string }>}
 *
 * At transfer: replace the measureText argument with a live fetch via the Power BI remote MCP
 * (powerbi-remote: getModelSchema + executeDax) for a named report/measure, and pass the
 * actual execution durationMs from the query result into stats.
 */
export function analyzeDax(measureText = '', stats = {}) {
  const text = String(measureText);
  const suggestions = RULES.filter(r => r.test.test(text)).map(({ pattern, suggestion }) => ({ pattern, suggestion }));
  if ((stats.durationMs ?? 0) >= 5000 && suggestions.length === 0) {
    suggestions.push({
      pattern: 'slow-no-obvious-cause',
      suggestion: `Measure runs in ${stats.durationMs} ms with no obvious anti-pattern — profile with Performance Analyzer / DAX Studio and check the storage-engine vs formula-engine split.`,
    });
  }
  return suggestions;
}

File: fabric-audit-agent-py/fabric_audit_agent/dax.py
"""DAX performance anti-pattern hints (rule-based text analysis). Port of ``core/dax.js``.

Heuristic — may have false positives; intended as hints, not verdicts. At deployment the
measure text can come from the Power BI remote MCP (getModelSchema + executeDax).
"""
import re

_RULES = [
    (re.compile(r"FILTER\s*\(\s*'?[A-Za-z0-9_ ]+'?\s*,", re.I), "filter-whole-table",
     "FILTER() over an entire table is slow — filter a single column (KEEPFILTERS or a boolean column filter) instead."),
    (re.compile(r"(SUMX|AVERAGEX|MAXX|MINX|COUNTX)[\s\S]*?(SUMX|AVERAGEX|MAXX|MINX|COUNTX)", re.I), "nested-iterators",
     "Possible nested iterators (X-functions inside X-functions). If nested, they evaluate row-by-row quadratically — precompute with a VAR or an aggregation table."),
    (re.compile(r"\bCALCULATE\b[\s\S]*\bCALCULATE\b", re.I), "repeated-calculate",
     "Repeated CALCULATE recomputes filter context — hoist shared expressions into VARs."),
    (re.compile(r"\bEARLIER\s*\(", re.I), "earlier",
     "EARLIER() signals legacy row-context nesting — refactor with VARs."),
    (re.compile(r"(?<![/:])[^/]/[^/]"), "raw-division",
     'Use DIVIDE(numerator, denominator) instead of "/" for safe, efficient division.'),
]


def analyze_dax(measure_text="", stats=None):
    stats = stats or {}
    text = str(measure_text)
    suggestions = [{"pattern": p, "suggestion": s} for (rx, p, s) in _RULES if rx.search(text)]
    if (stats.get("durationMs") or 0) >= 5000 and len(suggestions) == 0:
        suggestions.append({
            "pattern": "slow-no-obvious-cause",
            "suggestion": f"Measure runs in {stats.get('durationMs')} ms with no obvious anti-pattern — profile with Performance Analyzer / DAX Studio and check the storage-engine vs formula-engine split.",
        })
    return suggestions
